File: mobile/kmp/ui/src/commonMain/kotlin/com/atls/hyperion/ui/components/button/styles/appearance/Variants.kt
package com.atls.hyperion.ui.components.button.styles.appearance

import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import com.atls.hyperion.ui.theme.LocalHyperionColors

@Composable
fun ButtonAppearance.Companion.primary(): ButtonAppearance {
    val colors = LocalHyperionColors.current

    return ButtonAppearance(
        default = Colors.Solid(
            background = colors.action.base,
            content = colors.text.inverse,
            border = colors.action.base
        ),
        pressed = Colors.Solid(
            background = colors.action.pressed,
            content = colors.text.inverse,
            border = colors.action.pressed
        ),
        disabled = Colors.Solid(
            background = colors.action.disabled,
            content = colors.text.muted,
            border = colors.action.subtle
        ),
        focused = Colors.Solid(
            background = colors.action.base,
            content = colors.text.inverse,
            border = colors.action.pressed
        )
    )
}

@Composable
fun ButtonAppearance.Companion.secondary(): ButtonAppearance {
    val colors = LocalHyperionColors.current

    return ButtonAppearance(
        default = Colors.Solid(
            background = colors.action.subtle,
            content = colors.text.secondary,
            border = colors.surface.muted
        ),
        pressed = Colors.Solid(
            background = colors.text.secondary,
            content = colors.text.muted,
            border = colors.action.subtle
        ),
        disabled = Colors.Solid(
            background = colors.surface.muted,
            content = colors.text.tertiary,
            border = colors.surface.muted
        ),
        focused = Colors.Solid(
            background = colors.action.disabled,
            content = colors.text.secondary,
            border = colors.action.hover
        )
    )
}

@Composable
fun ButtonAppearance.Companion.ghost(): ButtonAppearance {
    val colors = LocalHyperionColors.current

    return ButtonAppearance(
        default = Colors.Solid(
            background = Color.Transparent,
            content = colors.text.secondary,
            border = Color.Transparent
        ),
        pressed = Colors.Solid(
            background = colors.action.subtle,
            content = colors.text.tertiary,
            border = colors.action.subtle
        ),
        disabled = Colors.Solid(
            background = Color.Transparent,
            content = colors.text.tertiary,
            border = Color.Transparent
        ),
        focused = Colors.Solid(
            background = Color.Transparent,
            content = colors.text.secondary,
            border = colors.action.disabled
        )
    )
}

File: ui-parts/button/src/styles/appearance/variants.ts
import type { Theme }                  from '@atls-ui/theme'

import type { ButtonAppearance }       from './interfaces.js'
import type { ButtonAppearanceStates } from './interfaces.js'
import type { ButtonAppearances }      from './interfaces.js'
import type { ButtonAppearanceName }   from './interfaces.js'

const variants = (theme: Theme): Record<ButtonAppearanceName, ButtonAppearanceStates> => ({
  primary: {
    default: {
      background: theme.colors.action.base,
      content: theme.colors.text.inverse,
      border: theme.colors.action.base,
    },
    hover: {
      background: theme.colors.action.hover,
      content: theme.colors.text.inverse,
      border: theme.colors.action.hover,
    },
    pressed: {
      background: theme.colors.action.pressed,
      content: theme.colors.text.inverse,
      border: theme.colors.action.pressed,
    },
    disabled: {
      background: theme.colors.action.disabled,
      content: theme.colors.text.muted,
      border: theme.colors.action.subtle,
    },
    focused: {
      background: theme.colors.action.base,
      content: theme.colors.text.inverse,
      border: theme.colors.action.pressed,
    },
  },
  secondary: {
    default: {
      background: theme.colors.action.subtle,
      content: theme.colors.text.secondary,
      border: theme.colors.surface.muted,
    },
    hover: {
      background: theme.colors.action.disabled,
      content: theme.colors.text.secondary,
      border: theme.colors.action.disabled,
    },
    pressed: {
      background: theme.colors.text.secondary,
      content: theme.colors.text.muted,
      border: theme.colors.action.subtle,
    },
    disabled: {
      background: theme.colors.surface.muted,
      content: theme.colors.text.tertiary,
      border: theme.colors.surface.muted,
    },
    focused: {
      background: theme.colors.action.disabled,
      content: theme.colors.text.secondary,
      border: theme.colors.action.hover,
    },
  },
  ghost: {
    default: {
      background: 'transparent',
      content: theme.colors.text.secondary,
      border: 'transparent',
    },
    hover: {
      background: theme.colors.surface.muted,
      content: theme.colors.text.secondary,
      border: theme.colors.surface.muted,
    },
    pressed: {
      background: theme.colors.action.subtle,
      content: theme.colors.text.tertiary,
      border: theme.colors.action.subtle,
    },
    disabled: {
      background: 'transparent',
      content: theme.colors.text.tertiary,
      border: 'transparent',
    },
    focused: {
      background: 'transparent',
      content: theme.colors.text.secondary,
      border: theme.colors.action.disabled,
    },
  },
})

export const buttonAppearances = {
  primary: 'primary',
  secondary: 'secondary',
  ghost: 'ghost',
} as const satisfies ButtonAppearances

export const resolveButtonAppearance = (
  appearance: ButtonAppearance,
  theme: Theme
): ButtonAppearanceStates =>
  typeof appearance === 'string' ? variants(theme)[appearance] : appearance
